text/0000-format-args-implicit-identifiers.md

```diff
@@ -0,0 +1,487 @@
+
+This RFC argues that `format_args!` should not attempt to expand any implicit named arguments if the macro is provided with an expression instead of a verbatim string literal.
+
+The following are motivations why this RFC argues this case:
+
+* This RFC's motivation for implicit named arguments is to give users a concise syntax for string formatting. When the format string is generated from some other expression this motivation for concise syntax is irrelevant.
+
+* The hygienic context of the string literal generated by the expansion is entirely dependent on the expression. For example, the string literal produced by the `concat!` macro resides in a separate hygienic context. In combination with implicit named arguments using hygiene inherited from the format string, this would lead to puzzling errors like the below:
+
+      error[E0425]: cannot find value `person` in this scope
+       --> scratch/test.rs:4:14
+        |
+        |     let person = "Charlie";
+      4 |     println!(concat!("hello {person}"));
+        |              ^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
+
+* The expression may expand to a format string which contains new identifiers not written by the users, bypassing macro hygiene in suprising ways. For example, if the `concat!` macro did not have the hygiene issue described above, it could be to "splice together" an implicit named argument like so:
+
+       let person = "Charlie";
+       println!(concat!("hello {p", "er", "son", "}"));
+
+   The RFC author argues that it appears highly undesirable that implicit capture of the `person` identifier should occur in this example given above.
+
+* Using the hygienic context of the format string for implicit named arguments can have potentially suprising results even just with `macro_rules!` macros.
+
+  For example, the RFC author found that with a proof-of-concept implementation of implicit named arguments the invocation below would print `"Snoopy"`:
+
+      const PERSON: &'static str = "Charlie";
+
+      fn main() {
+          macro_rules! bar {
+            () => { "{PERSON}" };
+          }
+
+          const PERSON: &'static str = "Snoopy";
+          println!(bar!());
+      }
+
+  However, by merely changing to `let` bindings and moving the `"Charlie"` declaration three lines down to be inside the `main()` function, as below, the invocation would instead print `"Charlie"`:
+
+      fn main() {
+          let person = "Charlie";
+          macro_rules! bar {
+              () => { "{person}" };
+          }
+
+          let person = "Snoopy";
+          println!(bar!());
+      }
+
+   While it can be argued that this example is very contrived, the RFC author believes that it is undesirable to add such subtle interactions to the `format_args!` family of macros.
+
+These appear to give strong motivation to disable implicit argument capture when `format_args!` expands an expression instead of a verbatim string literal.
```

Comment:
Let's make it an error instead, that's the most conservative variant.

Literal spans weren't significant for resolution previously (at least in built-in proc macros), so they were never really specified.
The literal returned by `concat!("a", "b")` could arguably have a call-site context, similarly to the result of `concat_idents!(a, b)`, so we may want to make things like `format!(concat!("foo", "{bar}"))` work eventually.

Reply:
The problem is that code compiles today, so making it an error is a breaking change.

Reply:
Which code compiles?
```rust
fn main() {
    let x = format!(concat!("{fo", "o}")); // error: there is no argument named `foo`
}
```

Reply:
Oh, whoops, sorry, mixed up which macros we were talking about.

This compiles in `panic!` due to the single-arg exception.

Reply:
> Let's make it an error instead, that's the most conservative variant.

Can you clarify a little what you'd like to see the error say? Disabling implicit named arguments would already cause the error you show above to be emitted (e.g. "there is no argument named `foo`")

Reply:
Well, something like "implicit named arguments are not supported in format strings expanded from macros"?

Reply:
I see. I'd also like to leave "no argument named foo" error untouched, but how about using a note. Something like the following:

    fn main() {
        let x = format!(concat!("{fo", "o}"));
        // ^~ error: there is no argument named `foo`
    }